### ISSUE TITLE
createComputePipelineState failed with E_INVALIDARG

### ISSUE THREAD

**Issue by serious-angel** [FransBouma/ShaderToggler#26]
Greetings! ✨

Thank you very much for the so awesome project!

---

May I ask regarding the following issue encountered?

I've just installed ReShade for `Need For Speed: Unbound`. In the result, launching of the title results in a crash with the error shown as in the preview attached (`E_INVALIDARG ("An invalid parameter was passed to the returning function.")`) during function execution `createComputePipelineState`.

![image](https://github.com/user-attachments/assets/466ba8ad-5613-4085-8883-f67b480cda91)

---

The steps followed to install ReShade were:
1. Download the latest installer version (`v6.3.1`) of ReShade with add-on support from the official resource;
2. While installing, chose the location of the `NeedForSpeedUnbound.exe` (inside `\Steam\steamapps\common\Excalibur\`);
3. Selected Direct 11/12 as the renderer;
4. Selected all the shaders to download;
5. Selected 2 add-on to install: `ShaderToggler` and `ReshadeEffectShaderToggler`.

Here, launching results in the crash mentioned.

---

With either the addon file `ShaderToggler.addon64` or `ReshadeEffectShaderToggler.addon64` located in the title directory and title re-laucnh, the error is shown.

With both addon files removal from the title directory and title re-launch, no error appears.

Please correct if mistaken, considering the error, is it possible the ReShade add-on is not compatible with the current driver in use where the function parameters does not match the function declaration?

---

What might it be or how would you try debugging the case case to find the root cause?

Best and kind regards 🪐

**Comment by serious-angel** [comment]
I got a suggestion in the Discord guild of the creator of mod [Camera Toolkit](<https://nfsmods.xyz/mod/4364>) for `Need for Speed: Unbound`. There was a member (`@mik_sharov`) who had no issues with both ReShade and the ShaderToggler add-on, and the latest video-game title installed, and they suggested to try out the same ReShade version as they have - `5.8.0`, and it worked! I've modified ReShade using the older installer, re-launched the title, terminated, added a single add-on file `ShaderToggler.addon64` of its latest version, launched again, and voila - few shaders could not compile, but there's no error, and the add-on is working like a charm! Freaking marvelous...

> Into the light may you fall
> Into the light may you follow
> Into the light may you know
> ~ Words Darker Than Their Wings - Alter Bridge

**Comment by FransBouma** [comment]
Strange, considering I use the latest reshade with shadertoggler all the time and no issues...

**Comment by serious-angel** [comment]
> Strange, considering I use the latest reshade with shadertoggler all the time and no issues...

Thank you for the response! Could it be the driver where ReShade behaves differently (e.g. changes the function declaration)?

**Comment by FransBouma** [comment]
Honestly, no idea...

**Comment by serious-angel** [comment]
> Honestly, no idea...

I haven't found any historical difference, too, for function in Nvidia driver changelogs and the repo as of now, except the following:

\- Commit [`2024-08-13T23:07:06.000Z`](<https://github.com/crosire/reshade/commit/ee2eef8bdf31cfe0ad3be8082d418a5656402f5c#diff-2125a9d47e5db47f56a7c50c42c6ec9111e64a9030fe0661b5d0eb5fbdb6e2f4R286>)
\- Commit [`2023-12-21T02:03:34.000Z`](<https://github.com/crosire/reshade/blame/acf87102daaa9b077847d28eb7ef802a9e1c85ac/source/d3d12/d3d12_device.cpp#L281>)

---

There's also a chance it's something related to the driver. I've found at least several mentions regarding it, but those are fairly dated:

> I fixed this, I uninstalled NVIDIA latest drivers (516.59v 06/28/2022) and installed previous version of drivers (516.40v 06/15/2022)
> [Source](<https://answers.ea.com/t5/Technical-Issues/-/m-p/11664938/highlight/true#M28556>)

> ...a DDU with my network disabled fixed it
> [Source](<https://redd.it/zcck3n>)

> Updating driver worked.
> [Source](<https://redd.it/13s5pkl>)

---

Interesting that the `NvAPI` function declaration differs from `DirectX 12` it seems:

```c
HRESULT CreateComputePipelineState(
  [in]  const D3D12_COMPUTE_PIPELINE_STATE_DESC *pDesc,
        REFIID                                  riid,
  [out] void                                    **ppPipelineState
);
```
[Source](<https://learn.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12device-createcomputepipelinestate>)

```c
NVAPI_INTERFACE NvAPI_D3D12_CreateComputePipelineState (
    __in ID3D12Device *                                 pDevice,
    __in const D3D12_COMPUTE_PIPELINE_STATE_DESC *      pPSODesc,
    NvU32 	                                        numExtensions,
    __in const NVAPI_D3D12_PSO_EXTENSION_DESC **        ppExtensions,
    __out ID3D12PipelineState **                        ppPSO 
)	
```
[Source](<https://docs.nvidia.com/gameworks/content/gameworkslibrary/coresdk/nvapi/group__dx.html#ga9a0848dc3f0024db2d66a40ce9359866>)

Related: [There are various other NVIDIA-specific pipeline state extensions that...](<https://developer.nvidia.com/blog/unlocking-gpu-intrinsics-in-hlsl/>)

**Comment by dakshcs** [comment]
bumping this thread

**Comment by FransBouma** [comment]
> bumping this thread

to accomplish that I jump at it right away? I have no repro nor any idea what it could be. So this won't be fixed from my side as I can't fix what I can't repro.

**Comment by dakshcs** [comment]
> > bumping this thread
> 
> to accomplish that I jump at it right away? I have no repro nor any idea what it could be. So this won't be fixed from my side as I can't fix what I can't repro.

no not right away, just like on your to-do list someday.

I'll get you a screen recording and logs the next time I play unbound. (it is interesting because i have an AMD card from 2018 instead of an RTX card)

also sorry if i seemed a bit rude.

**Comment by FransBouma** [comment]
Well, as I said, I can't reproduce it locally with any other game (and I don't have this particular game, nor the time to look into this) so investing time on your side in this might not result in the result you're hoping for. For this particular game, it might be better to roll back to reshade 5.8 as that seems to work according to @serious-angel

**Comment by crosire** [comment]
This is a problem caused by ReShade and fixed by https://github.com/crosire/reshade/commit/c463528a4893e2b577227d36c93f5b3d68eebe09.

**Comment by FransBouma** [comment]
> This is a problem caused by ReShade and fixed by [crosire/reshade@c463528](https://github.com/crosire/reshade/commit/c463528a4893e2b577227d36c93f5b3d68eebe09).

Awesome, thanks man :)

**Comment by dakshcs** [comment]
> This is a problem caused by ReShade and fixed by [crosire/reshade@c463528](https://github.com/crosire/reshade/commit/c463528a4893e2b577227d36c93f5b3d68eebe09).

thanks so much
